### PR TITLE
Lily/generator

### PR DESCRIPTION
**Pull request by lhz1029**

### CODE REVIEW

File: finetune/input_pipeline.py [hunk 1]
```diff
@@ -187,18 +193,21 @@ def _dataset_with_targets(self, Xs, Y, train, context=None, update_hook=None):
...
+        dataset_gen = Dataset.from_generator(
             lambda: self.wrap_tqdm(dataset_encoded(), train, update_hook=update_hook), *shape_def
         )
+        return dataset_gen
 
     def _dataset_without_targets(self, Xs, train, context=None, update_hook=None):
         if context is not None:
             # we assume that X must have known length if we also provide context so this is safe
-            if callable(Xs):
+            if callable(Xs) and callable(context):
```
Comment (madisonmay) [inline finetune/input_pipeline.py:204]: Eventually (before merge) we'll also need this to handle the case where context is None

File: finetune/nn/auxiliary.py [hunk 2]
```diff
@@ -121,7 +125,22 @@ def dense_with_custom_init(input_tensor,
             return z
 
     else:
-        return tf.layers.dense(input_tensor, output_dim, activation=activation, name=name, kernel_initializer=kernel_initializer)
+        if transpose_b:
```
Comment (madisonmay) [inline finetune/nn/auxiliary.py:128]: Could we always use the path that doesn't use tf.layers.dense?  And only conditionally transpose?

**Review by madisonmay — APPROVED** [review]

File: finetune/nn/auxiliary.py (hunk 2 above)
Comment (lhz1029) [reply finetune/nn/auxiliary.py:128]: Yeah I could do that. Just need to do a bit of testing to make sure the naming is correct

**Review by benleetownsend — APPROVED** [review]

**Review by Concomitant — COMMENTED** [review]
LGTM!

**Review by Concomitant — APPROVED** [review]

File: finetune/input_pipeline.py (hunk 1 above)
Comment (lhz1029) [reply finetune/input_pipeline.py:204]: it's handled in the other if statement

**Pull request merged**